lucene/core/src/java/org/apache/lucene/search/FilteringNumericComparator.java

```diff
@@ -0,0 +1,329 @@
+    protected DocIdSetIterator competitiveIterator = null;
+    private long iteratorCost = 0;
+
+    public FilteringNumericComparator(NumericComparator<T> in, boolean reverse, int bytesCount) {
+        super(in);
+        this.field = in.field;
+        this.bytesCount = bytesCount;
+        this.reverse = reverse;
+        minValueAsBytes = new byte[bytesCount];
+        maxValueAsBytes = new byte[bytesCount];
+        if (reverse) {
+            minValueExist = true;
+        } else {
+            maxValueExist = true;
+        }
+    }
+
+    /**
+     * Returns an iterator over competitive documents
+     */
+    @Override
+    public DocIdSetIterator competitiveIterator() {
+        if (competitiveIterator == null) return null;
+        return new DocIdSetIterator() {
+            private int doc;
+            @Override
+            public int nextDoc() throws IOException {
+                return doc = competitiveIterator.nextDoc();
+            }
+
+            @Override
+            public int docID() {
+                return doc;
+            }
+
+            @Override
+            public long cost() {
+                return competitiveIterator.cost();
+            }
+
+            @Override
+            public int advance(int target) throws IOException {
+                return doc = competitiveIterator.advance(target);
+            }
+        };
+    }
+
+    @Override
+    public void setCanUpdateIterator() throws IOException {
+        this.canUpdateIterator = true;
+        // for the 1st time queue becomes full and hitsThreshold is reached
+        // we can start updating competitive iterator
+        updateCompetitiveIterator();
+    }
+
+    @Override
+    public void setTopValue(T value) {
+        hasTopValue = true;
+        if (reverse) {
+            maxValueExist = true;
+        } else {
+            minValueExist = true;
+        }
+        in.setTopValue(value);
+    }
+
+    @Override
+    public void setBottom(int slot) throws IOException {
+        ((NumericComparator) in).setBottom(slot);
+        updateCompetitiveIterator(); // update an iterator if we set a new bottom
+    }
+
+    @Override
+    public int compareBottom(int doc) throws IOException {
+        return ((NumericComparator) in).compareBottom(doc);
+    }
+
+    @Override
+    public int compareTop(int doc) throws IOException {
+        return ((NumericComparator) in).compareTop(doc);
+    }
+
+    @Override
+    public void copy(int slot, int doc) throws IOException {
+        ((NumericComparator) in).copy(slot, doc);
+    }
+
+    @Override
+    public void setScorer(Scorable scorer) throws IOException {
+        ((NumericComparator) in).setScorer(scorer);
+        if (scorer instanceof Scorer) {
+            iteratorCost = ((Scorer) scorer).iterator().cost(); // starting iterator cost is the scorer's cost
+            updateCompetitiveIterator(); // update an iterator when we have a new segment
+        }
```

Comment:
Let's open a separate issue about adding `Scorable.cost()` to make such optimizations easier?

Reply:
Opened  [LUCENE-9352](https://issues.apache.org/jira/browse/LUCENE-9352)